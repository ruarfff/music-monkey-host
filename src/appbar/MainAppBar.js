import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from 'material-ui/styles'
import AppBar from 'material-ui/AppBar'
import Toolbar from 'material-ui/Toolbar'
import Typography from 'material-ui/Typography'
import IconButton from 'material-ui/IconButton'
import Menu, { MenuItem } from 'material-ui/Menu'
import Avatar from 'material-ui/Avatar';

const styles = {
  root: {
    flexGrow: 1
  },
  title: {
    flex: 1
  },
  menuButton: {
    marginLeft: -12,
    marginRight: 20
  }
}

class MainAppBar extends React.Component {
  state = {
    anchorEl: null
  }

  handleMenu = event => {
    this.setState({ anchorEl: event.currentTarget })
  }

  handleClose = () => {
    this.setState({ anchorEl: null })
  }

  render() {
    const { classes, user } = this.props
    const { anchorEl } = this.state
    const open = Boolean(anchorEl)

    return (
      <div className={classes.root}>
        <AppBar position="static">
          <Toolbar>
            <Typography
              align="center"
              variant="title"
              color="inherit"
              className={classes.title}
            >
              MusicMonkey
            </Typography>

            <div>
              <IconButton
                aria-owns={open ? 'menu-appbar' : null}
                aria-haspopup="true"
                onClick={this.handleMenu}
                color="inherit"
              >
                <Avatar alt="user profile" src={user.images[0].url} />
              </IconButton>
              <Menu
                id="menu-appbar"
                anchorEl={anchorEl}
                anchorOrigin={{
                  vertical: 'top',
                  horizontal: 'right'
                }}
                transformOrigin={{
                  vertical: 'top',
                  horizontal: 'right'
                }}
                open={open}
                onClose={this.handleClose}
              >
                <MenuItem onClick={this.handleClose}>Profile</MenuItem>
                <MenuItem onClick={this.handleClose}>My account</MenuItem>
              </Menu>
            </div>
          </Toolbar>
        </AppBar>
      </div>
    )
  }
}

MainAppBar.propTypes = {
  classes: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired
}

export default withStyles(styles)(MainAppBar)
