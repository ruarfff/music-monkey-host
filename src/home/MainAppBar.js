import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from 'material-ui/styles'
import classNames from 'classnames'
import MenuIcon from '@material-ui/icons/Menu'
import AppBar from 'material-ui/AppBar'
import Toolbar from 'material-ui/Toolbar'
import Typography from 'material-ui/Typography'
import IconButton from 'material-ui/IconButton'
import Menu, { MenuItem } from 'material-ui/Menu'
import Avatar from 'material-ui/Avatar'
import AccountCircle from '@material-ui/icons/AccountCircle'
import { drawerWidth } from './homeConstants'

const styles = theme => ({
  title: {
    flex: 1
  },
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
    transition: theme.transitions.create(['width', 'margin'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen
    })
  },
  appBarShift: {
    marginLeft: drawerWidth,
    width: `calc(100% - ${drawerWidth}px)`,
    transition: theme.transitions.create(['width', 'margin'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen
    })
  },
  menuButton: {
    marginLeft: 12,
    marginRight: 36
  },
  hide: {
    display: 'none'
  },
  profile: {
    marginRight: 12
  },
  profileSideOpen: {
    marginRight: -12
  }
})

class MainAppBar extends React.Component {
  state = {
    anchorEl: null
  }

  handleMenu = event => {
    this.setState({ anchorEl: event.currentTarget })
  }

  handleClose = () => {
    this.setState({ anchorEl: null })
  }

  render() {
    const { classes, user, openSidebar } = this.props
    const { sidebarIsOpen } = this.props.home
    const { anchorEl } = this.state
    const open = Boolean(anchorEl)
    const userHasProfileImage = user.images && user.images.length > 0

    const menuButton = (
      <IconButton
        color="inherit"
        aria-label="open drawer"
        onClick={openSidebar}
        className={classNames(
          classes.menuButton,
          sidebarIsOpen && classes.hide
        )}
      >
        <MenuIcon />
      </IconButton>
    )

    const profilePic = (
      <div
        className={classNames(
          classes.profile,
          sidebarIsOpen && classes.profileSideOpen
        )}
      >
        <IconButton
          aria-owns={open ? 'menu-appbar' : null}
          aria-haspopup="true"
          onClick={this.handleMenu}
          color="inherit"
        >
          {userHasProfileImage ? (
            <Avatar alt="user profile" src={user.images[0].url} />
          ) : (
            <AccountCircle />
          )}
        </IconButton>
        <Menu
          id="menu-appbar"
          anchorEl={anchorEl}
          anchorOrigin={{
            vertical: 'top',
            horizontal: 'right'
          }}
          transformOrigin={{
            vertical: 'top',
            horizontal: 'right'
          }}
          open={open}
          onClose={this.handleClose}
        >
          <MenuItem onClick={this.handleClose}>My account</MenuItem>
          <MenuItem onClick={this.handleClose}>Logout</MenuItem>
        </Menu>
      </div>
    )

    return (
      <AppBar
        position="absolute"
        className={classNames(
          classes.appBar,
          sidebarIsOpen && classes.appBarShift
        )}
      >
        <Toolbar disableGutters={!sidebarIsOpen}>
          {menuButton}
          <Typography variant="title" color="inherit" className={classes.title}>
            MusicMonkey
          </Typography>
          {profilePic}
        </Toolbar>
      </AppBar>
    )
  }
}

MainAppBar.propTypes = {
  classes: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired,
  openSidebar: PropTypes.func.isRequired,
  home: PropTypes.object.isRequired
}

export default withStyles(styles)(MainAppBar)
