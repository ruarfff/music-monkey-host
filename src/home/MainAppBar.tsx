import AppBar from '@material-ui/core/AppBar/AppBar'
import Avatar from '@material-ui/core/Avatar/Avatar'
import Badge from '@material-ui/core/Badge'
import Button from '@material-ui/core/Button'
import IconButton from '@material-ui/core/IconButton/IconButton'
import Menu from '@material-ui/core/Menu/Menu'
import MenuItem from '@material-ui/core/MenuItem/MenuItem'
import { withStyles, WithStyles } from '@material-ui/core/styles'
import Toolbar from '@material-ui/core/Toolbar/Toolbar'
import Typography from '@material-ui/core/Typography/Typography'
import AccountCircle from '@material-ui/icons/AccountCircle'
import NotificationsIcon from '@material-ui/icons/Notifications'
import * as React from 'react'
import { Link } from 'react-router-dom'
import eventIcon from '../assets/event-icon.svg'
import NotificationPopup from '../components/NotificationPopup/NotificationPopup'
import IEvent from '../event/IEvent'
import IAction from '../IAction'
import { onRsvpSaved } from '../notification'
import {
  INotification,
  INotificationState
} from '../notification/notificationInitialState'
import IUser from '../user/IUser'

const decorate = withStyles(({ transitions, zIndex }) => ({
  root: {},
  appBar: {
    transition: transitions.create(['width', 'margin'], {
      duration: transitions.duration.leavingScreen,
      easing: transitions.easing.sharp
    }),
    zIndex: zIndex.drawer + 1,
    position: 'relative',
    boxShadow: 'none',
    backgroundColor: 'white',
    color: 'black',
    paddingTop: '30px'
  },
  hide: {
    display: 'none'
  },
  menuButton: {
    marginLeft: 12,
    marginRight: 36
  },
  profile: {
    marginRight: 12,
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    flexDirection: 'row'
  },
  title: {
    flex: 1,
    cursor: 'pointer',
    fontSize: '34px',
    lineHeight: '40px',
    padding: '16px'
  },
  addEventBtn: {
    textTransform: 'uppercase',
    padding: '8px 18px'
  },
  notification: {
    color: '#979797'
  },
  userName: {
    color: '#979797'
  },
  imageInButton: {
    marginRight: '5px'
  }
}))

interface IMainAppBarProps {
  user: IUser
  location: string
  event: IEvent
  notification: INotificationState
  logout(): IAction
  getNotifications(id: string): IAction
  handleTitleClicked(): void
  updateNotification(notification: INotification): IAction
}

class MainAppBar extends React.Component<IMainAppBarProps & WithStyles> {
  public state = {
    notificationAnchor: undefined,
    anchorEl: undefined,
    showNotification: false
  }

  public handleMenu = (event: any) => {
    this.setState({ anchorEl: event.currentTarget })
  }

  public handleClose = () => {
    this.setState({ anchorEl: undefined })
  }

  public componentDidMount() {
    const { user, getNotifications } = this.props
    onRsvpSaved(user.userId, getNotifications(user.userId))
  }

  public menuName = (history: string) => {
    const { event } = this.props

    switch (history) {
      case '/':
        return 'Dashboard'
      case '/create-event':
        return 'Create Event'
      case '/all-events':
        return 'Events'
      case '/upcoming-events':
        return 'Events'
      case '/past-events':
        return 'Events'
      case `/events/${event && event.eventId}`:
        return event.name
      case '/account':
        return 'Account Settings'
      case '/insights':
        return 'Insights'
      default:
        return 'Dashboard'
    }
  }

  public toggleNotification = (event?: any) => {
    if (event) {
      this.setState({
        showNotification: !this.state.showNotification,
        notificationAnchor: event.currentTarget
      })
    }
  }

  public render() {
    const {
      classes,
      user,
      location,
      handleTitleClicked,
      notification,
      updateNotification
    } = this.props
    const { anchorEl } = this.state
    const open = Boolean(anchorEl)
    const userHasProfileImage = !!user && !!user.image
    const userHasName = !!user && !!user.displayName

    const unreadNotifications = notification.notifications.filter(
      n => n.status === 'Unread'
    )

    const profilePic = (
      <div className={classes.profile}>
        {location !== '/create-event' && (
          <Link to="/create-event" className="Home-create-event-link">
            <Button
              variant="contained"
              size="small"
              color='secondary'
              className={classes.addEventBtn}
            >
              <img className={classes.imageInButton} src={eventIcon} alt="" />
              Create new event
            </Button>
          </Link>
        )}
        <IconButton
          aria-owns={this.state.showNotification ? 'menu-notification' : undefined}
          aria-haspopup="true"
          color="inherit"
          onClick={this.toggleNotification}
        >
          <Badge
            className={classes.notification}
            badgeContent={unreadNotifications ? unreadNotifications.length : 0}
            color="primary"
          >
            <NotificationsIcon />
          </Badge>
          <NotificationPopup
            notificationAnchor={this.state.notificationAnchor}
            toggleNotification={this.toggleNotification}
            showNotification={this.state.showNotification}
            notifications={notification.notifications}
            updateNotification={updateNotification}
          />
        </IconButton>

        <IconButton
          aria-owns={open ? 'menu-appbar' : undefined}
          aria-haspopup="true"
          onClick={this.handleMenu}
          color="inherit"
        >
          {userHasProfileImage ? (
            <Avatar alt="user profile" src={user.image} />
          ) : (
            <AccountCircle />
          )}
        </IconButton>
        {userHasName ? (
          <Typography className={classes.userName} onClick={handleTitleClicked}>
            {user.displayName}
          </Typography>
        ) : (
          <Typography className={classes.userName} onClick={handleTitleClicked}>
            Name
          </Typography>
        )}
        <Menu
          id="menu-appbar"
          anchorEl={anchorEl}
          anchorOrigin={{
            horizontal: 'right',
            vertical: 'top'
          }}
          transformOrigin={{
            horizontal: 'right',
            vertical: 'top'
          }}
          open={open}
          onClose={this.handleClose}
        >
          <Link to={'/account'}>
            <MenuItem onClick={this.handleClose}>My Account</MenuItem>
          </Link>
          <MenuItem onClick={this.props.logout}>Logout</MenuItem>
        </Menu>
      </div>
    )

    return (
      <AppBar className={classes.appBar}>
        <Toolbar>
          <Typography
            variant="h6"
            color="inherit"
            className={classes.title}
            onClick={handleTitleClicked}
          >
            {this.menuName(location)}
          </Typography>
          {profilePic}
        </Toolbar>
      </AppBar>
    )
  }
}

export default decorate(MainAppBar)
