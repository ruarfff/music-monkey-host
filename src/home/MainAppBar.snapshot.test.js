import React from 'react'
import renderer from 'react-test-renderer'
import MainAppBar from './MainAppBar'

describe('MainAppBar Snapshot', () => {
  it('has not changed unexpectedly', () => {
    const user = {}
    const classes = {}
    const home = {}
    const openSidebar = () => {}
    const logout = () => {}
    const component = renderer.create(
      <MainAppBar
        user={user}
        classes={classes}
        home={home}
        openSidebar={openSidebar}
        logout={logout}
      />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
