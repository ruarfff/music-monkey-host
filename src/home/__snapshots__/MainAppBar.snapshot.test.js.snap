// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainAppBar Snapshot has not changed unexpectedly 1`] = `
<header
  className="MuiPaper-root-16 MuiPaper-elevation4-22 MuiAppBar-root-8 MuiAppBar-positionAbsolute-10 MuiAppBar-colorPrimary-14 MainAppBar-appBar-2"
>
  <div
    className="MuiToolbar-root-43"
  >
    <button
      aria-label="open drawer"
      className="MuiButtonBase-root-51 MuiIconButton-root-45 MuiIconButton-colorInherit-46 MainAppBar-menuButton-4"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiIconButton-label-50"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-53"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <g>
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </g>
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root-59"
      />
    </button>
    <h2
      className="MuiTypography-root-66 MuiTypography-title-72 MuiTypography-colorInherit-85 MainAppBar-title-1"
    >
      MusicMonkey
    </h2>
    <div
      className="MainAppBar-profile-6"
    >
      <button
        aria-haspopup="true"
        aria-owns={null}
        className="MuiButtonBase-root-51 MuiIconButton-root-45 MuiIconButton-colorInherit-46"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-50"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-53"
            color={undefined}
            focusable="false"
            viewBox="0 0 24 24"
          >
            <g>
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </g>
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-59"
        />
      </button>
    </div>
  </div>
</header>
`;
